application/pages/modplatform/ftb/FtbModel.cpp

```diff
@@ -0,0 +1,250 @@
+}
+
+void ListModel::getLogo(const QString &logo, const QString &logoUrl, LogoCallback callback)
+{
+    if(m_logoMap.contains(logo))
+    {
+        callback(ENV.metacache()->resolveEntry("ModpacksCHPacks", QString("logos/%1").arg(logo.section(".", 0, 0)))->getFullPath());
+    }
+    else
+    {
+        requestLogo(logo, logoUrl);
+    }
+}
+
+void ListModel::searchWithTerm(const QString &term)
+{
+    if(currentSearchTerm == term) {
+        return;
+    }
+    currentSearchTerm = term;
+    if(jobPtr) {
+        jobPtr->abort();
+        searchState = ResetRequested;
+        return;
+    }
+    else {
+        beginResetModel();
+        modpacks.clear();
+        endResetModel();
+        searchState = None;
+    }
+    performPaginatedSearch();
+}
+
+void ListModel::searchRequestFinished()
+{
+    jobPtr.reset();
+    remainingPacks.clear();
+
+    QJsonParseError parse_error;
+    QJsonDocument doc = QJsonDocument::fromJson(response, &parse_error);
+    if(parse_error.error != QJsonParseError::NoError) {
+        qWarning() << "Error while parsing JSON response from FTB at " << parse_error.offset << " reason: " << parse_error.errorString();
+        qWarning() << response;
+        return;
+    }
+
+    auto packs = doc.object().value("packs").toArray();
+    for(auto pack : packs) {
+        auto packId = pack.toInt();
+        remainingPacks.append(packId);
+    }
+
+    if(!remainingPacks.isEmpty()) {
+        currentPack = remainingPacks.at(0);
+        requestPack();
+    }
+}
+
+void ListModel::searchRequestFailed(QString reason)
+{
+    jobPtr.reset();
+    remainingPacks.clear();
+
+    if(searchState == ResetRequested) {
+        beginResetModel();
+        modpacks.clear();
+        endResetModel();
+
+        performPaginatedSearch();
+    } else {
+        searchState = Finished;
+    }
+}
+
+void ListModel::requestPack()
+{
+    auto *netJob = new NetJob("Ftb::Search");
+    auto searchUrl = QString(BuildConfig.MODPACKSCH_API_BASE_URL + "public/modpack/%1")
+            .arg(currentPack);
+    netJob->addNetAction(Net::Download::makeByteArray(QUrl(searchUrl), &response));
+    jobPtr = netJob;
+    jobPtr->start();
+
+    QObject::connect(netJob, &NetJob::succeeded, this, &ListModel::packRequestFinished);
+    QObject::connect(netJob, &NetJob::failed, this, &ListModel::packRequestFailed);
+}
+
+void ListModel::packRequestFinished()
+{
+    jobPtr.reset();
+    remainingPacks.removeOne(currentPack);
+
+    QJsonParseError parse_error;
+    QJsonDocument doc = QJsonDocument::fromJson(response, &parse_error);
+    if(parse_error.error != QJsonParseError::NoError) {
+        qWarning() << "Error while parsing JSON response from FTB at " << parse_error.offset << " reason: " << parse_error.errorString();
+        qWarning() << response;
+        return;
+    }
+
+    auto obj = doc.object();
+
+    ModpacksCH::Modpack pack;
+    ModpacksCH::loadModpack(pack, obj);
```

Comment:
Catch parsing errors here. Basically all of those `require*` JSON calls will throw if there's any problem. Uncaught exceptions lead to a crash.